### PR TITLE
Add modal transition systems and associated algorithms to automatalib

### PR DESCRIPTION
**Pull request by Conturing**

### CODE REVIEW

File: api/src/main/java/net/automatalib/ts/modal/ModalTransitionSystem.java [hunk 1]
```diff
@@ -15,46 +15,48 @@
...
+import net.automatalib.automata.visualization.MTSVisualizationHelper;
 import net.automatalib.graphs.FiniteLTS;
-import net.automatalib.visualization.DelegateVisualizationHelper;
+import net.automatalib.graphs.UniversalGraph;
+import net.automatalib.ts.modal.transitions.ModalEdgeProperty;
 import net.automatalib.visualization.VisualizationHelper;
 
 public interface ModalTransitionSystem<S, I, T, TP extends ModalEdgeProperty>
-        extends UniversalAutomaton<S, I, T, Void, TP>, FiniteLTS<S, T, I>, InputAlphabetHolder<I> {
+        extends UniversalFiniteAlphabetAutomaton<S, I, T, Void, TP>, FiniteLTS<S, T, I>, InputAlphabetHolder<I> {
```
Comment (mtf90) [inline api/src/main/java/net/automatalib/ts/modal/ModalTransitionSystem.java:33]: Do MTSs by definition have a finite set of states and inputs, or would it make sense (on an interface level) to define MTSs with unbounded states or inputs?

File: api/src/main/java/net/automatalib/automata/visualization/MMCVisualizationHelper.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+            }
+            else if (7 <= memberId && memberId <= 13) {
+                properties.put("colorscheme", "ylorrd9");
+                properties.put(EdgeAttrs.COLOR, String.valueOf(16 - memberId));
+            }
+            else if (14 <= memberId) {
+                properties.put("colorscheme", "rdpu9");
+                properties.put(EdgeAttrs.COLOR, String.valueOf(9 - (memberId % 7)));
+            }
+        }
```
Comment (mtf90) [inline api/src/main/java/net/automatalib/automata/visualization/MMCVisualizationHelper.java:48]: What purpose does the memberId have? Is it just for clustering? I think having it as a semantic property is fine, but things like `colorscheme` seems very DOT-specific (whereas the `VisualizationHelper` is intended to be visualization-agnostic) and magic numbers are a code smell. This kind of specific visualization would probably better fit in you application code.

File: api/src/main/java/net/automatalib/ts/modal/transitions/GroupMemberEdge.java [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package net.automatalib.ts.modal.transitions;
+
+public interface GroupMemberEdge<G> {
+
+    G getGroup();
```
Comment (mtf90) [inline api/src/main/java/net/automatalib/ts/modal/transitions/GroupMemberEdge.java:5]: Currently, I don't see this method used and at the only point where a `GroupMemberEdge` is used (`MMCVisualizationHelper`) it is used without a generic type parameter. I assume it is intended to associate complex objects with specific group member (ids) but wouldn't this fit better in an external mapping?

File: api/src/main/java/net/automatalib/ts/modal/transitions/MutableGroupMemberEdge.java [hunk 4]
```diff
@@ -0,0 +1,8 @@
+package net.automatalib.ts.modal.transitions;
+
+public interface MutableGroupMemberEdge<G> extends GroupMemberEdge<G> {
+
+    void setGroup(G group);
```
Comment (mtf90) [inline api/src/main/java/net/automatalib/ts/modal/transitions/MutableGroupMemberEdge.java:5]: See `GroupMemberEdge`.

File: pom.xml [hunk 5]
```diff
@@ -75,6 +75,28 @@ limitations under the License.
                 <role>Developer</role>
             </roles>
         </contributor>
+        <contributor>
```
Comment (mtf90) [inline pom.xml:78]: Marc doesn't want to be added? :)

File: serialization/dot/src/main/java/net/automatalib/serialization/dot/DOTMutableAutomatonParser.java [hunk 6]
```diff
@@ -117,7 +117,28 @@ public DOTMutableAutomatonParser(AutomatonCreator<A, I> creator,
     private <S> void parseNodesAndEdges(InternalDOTParser parser, MutableAutomaton<S, I, ?, SP, TP> automaton) {
         final Map<String, S> stateMap = Maps.newHashMapWithExpectedSize(parser.getNodes().size());
 
-        for (Node node : parser.getNodes()) {
```
Comment (mtf90) [inline serialization/dot/src/main/java/net/automatalib/serialization/dot/DOTMutableAutomatonParser.java:120]: I'll rework this as discussed

File: serialization/dot/src/main/java/net/automatalib/serialization/dot/GraphDOT.java [hunk 7]
```diff
@@ -131,8 +135,16 @@ public static void write(GraphViewable gv, Appendable a) throws IOException {
      * @throws IOException
      *         if writing to {@code a} fails.
      */
-    public static <N, E> void write(Graph<N, E> graph, Appendable a) throws IOException {
-        writeRaw(graph, a, toDOTVisualizationHelper(graph.getVisualizationHelper()));
+    public static <N, E> void write(Graph<N, E> graph, Appendable a, boolean startHasNewID) throws IOException {
```
Comment (mtf90) [inline serialization/dot/src/main/java/net/automatalib/serialization/dot/GraphDOT.java:138]: I'll rework this as discussed

File: serialization/dot/src/test/java/net/automatalib/serialization/dot/DOTDeserializationTest.java [hunk 8]
```diff
@@ -58,7 +62,7 @@ public void testRegularDFADeserialization() throws IOException {
                 DOTParsers.dfa().readModel(DOTSerializationUtil.getResource(DOTSerializationUtil.DFA_RESOURCE)).
                         model;
 
-        Assert.assertTrue(Automata.testEquivalence(dfa, parsed, dfa.getInputAlphabet()));
+        checkIsomorphism(dfa, parsed, dfa.getInputAlphabet());
```
Comment (mtf90) [inline serialization/dot/src/test/java/net/automatalib/serialization/dot/DOTDeserializationTest.java:65]: What does `checkIsomorphism` do differently than `Automata.testEquivalence` (other than being applicable to non-deterministic systems)? Why change tests for code you haven't (or shouldn't have) touched?

File: serialization/dot/src/test/java/net/automatalib/serialization/dot/DOTDeserializationTest.java [hunk 9]
```diff
@@ -160,22 +188,77 @@ public void doNotCloseInputStreamTest() throws IOException {
...
-                                                        .create();
-        // @formatter:on
+        final CompactDFA<String> dfa = new CompactDFA<>(Alphabets.closedCharStringRange('a', 'b'));
+        final Integer s0 = dfa.addInitialState();
+        final Integer s1 = dfa.addState(true);
+
+        dfa.addTransition(s0, "a", s1);
+        dfa.addTransition(s0, "b", s1);
+        dfa.addTransition(s1, "a", s0);
+        dfa.addTransition(s1, "b", s0);
```
Comment (mtf90) [inline serialization/dot/src/test/java/net/automatalib/serialization/dot/DOTDeserializationTest.java:198]: Same question as above: Why refactor unrelated code?

File: util/src/main/java/net/automatalib/util/ts/modal/MCUtil.java [hunk 10]
```diff
@@ -0,0 +1,421 @@
...
+
+public class MCUtil {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(SystemComponent.class);
+
+    private MCUtil() {
+        // prevent instantiation
+    }
+
+    public static CompactMC<String> loadMCFromPath(String path) throws IOException {
```
Comment (mtf90) [inline util/src/main/java/net/automatalib/util/ts/modal/MCUtil.java:59]: As discussed, some of the file-loading code may be better located in your application code.

File: api/src/main/java/net/automatalib/ts/modal/ModalTransitionSystem.java (hunk 1 above)
Comment (Conturing) [reply api/src/main/java/net/automatalib/ts/modal/ModalTransitionSystem.java:33]: Larsen (1990) defined MTSs with possible unbounded sets of states and inputs, but the transition relation should be image-finite.

File: util/src/test/java/net/automatalib/util/ts/modal/regression/RegressionTests.java [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+        CompactDFA<String> greenLanguage = (CompactDFA<String>) MCUtil.greenContextLanguage(instance.modal_contract);
+        CompactMTS<String> greenContext = MCUtil.greenContextComponent(greenLanguage,
+                new CompactMTS.Creator<String>(),
+                instance.modal_contract.getCommunicationAlphabet(),
+                () -> new ModalEdgePropertyImpl(ModalEdgeProperty.ModalType.MAY),
+                () -> new ModalEdgePropertyImpl(ModalEdgeProperty.ModalType.MUST));
+
+        CompactMTS<String> context = MTSUtil.conjunction(greenContext, redContext);
+
+        MTSUtil.saveMTSToPath(context, "src/test/resources/phil3-actual/"+ testCase.context);
```
Comment (mtf90) [inline util/src/test/java/net/automatalib/util/ts/modal/regression/RegressionTests.java:175]: Do you really want to write a file to a source directory? This makes repeated builds (even with `clean`) not produce the same artifacts. Is there any specific reason you need to write this (or any) file? I haven't seen it being read anywhere.

File: serialization/dot/src/test/java/net/automatalib/serialization/dot/DOTDeserializationTest.java (hunk 8 above)
Comment (Conturing) [reply serialization/dot/src/test/java/net/automatalib/serialization/dot/DOTDeserializationTest.java:65]: Not my change, see commit a3a967d

File: util/src/test/java/net/automatalib/util/ts/modal/regression/RegressionTests.java (hunk 11 above)
Comment (Conturing) [reply util/src/test/java/net/automatalib/util/ts/modal/regression/RegressionTests.java:175]: You are right, the source directory was a bad choice. It should have been a separate resource directory. The idea is to log (intermediate) graphs since the assertions currently do not provide sufficient error information on failure.

File: serialization/dot/src/test/java/net/automatalib/serialization/dot/DOTDeserializationTest.java (hunk 9 above)
Comment (Conturing) [reply serialization/dot/src/test/java/net/automatalib/serialization/dot/DOTDeserializationTest.java:198]: See response above

File: util/pom.xml [hunk 12]
```diff
@@ -52,6 +52,10 @@ limitations under the License.
             <groupId>net.automatalib</groupId>
             <artifactId>automata-core</artifactId>
         </dependency>
+        <dependency>
+            <groupId>net.automatalib</groupId>
+            <artifactId>automata-serialization-dot</artifactId>
```
Comment (mtf90) [inline util/pom.xml:57]: A runtime or test dependency?

File: api/src/main/java/net/automatalib/automata/visualization/MMCVisualizationHelper.java (hunk 2 above)
Comment (Conturing) [reply api/src/main/java/net/automatalib/automata/visualization/MMCVisualizationHelper.java:48]: It is, indeed, a form of clustering. However, in our use case these ids are an intrinsic property of the edges. It is important that the mapping edge -> id is available and update when using this graph, esp. in contexts like subgraphs or copies. Therefore it seemed appropriate to store this information inside the graph itself.
As for the colorschemes, we use them to visualize the clusters depending on the memberId and the color. The magic numbers are used to ensure visually distinguishable colors. When the `VisualizationHelper` is not suitable for this tasks, can you extract the code? Or should it be client specific?

File: api/src/main/java/net/automatalib/ts/modal/transitions/GroupMemberEdge.java (hunk 3 above)
Comment (Conturing) [reply api/src/main/java/net/automatalib/ts/modal/transitions/GroupMemberEdge.java:5]: At the time it seemed to be a more coherent and general concept than just defining an interface with memberIds. Currently, the group is always equivalent to the color attribute and therefore not essential.

**Review by mtf90 — CHANGES_REQUESTED** [review]
Most of the code looks fine. The `automata-serialization-dot` module currently doesn't pass the tests for me. This seems to be related to some changes regarding the rendering of the initial node. As I wanted to change some of the DOT serialization code anyway, I can look into it while I am at it, but double-checking that this is the only issue would be nice.

Other than that, there are only a few places where I had questions regarding design decisions. I annotated the corresponding lines and would appreciate some clarification.

Expect some refactorings and renamings (things like package names, removal of file loading utilities, etc.) but most of the features should remain as-is. Some of the (to be removed) utility functions can be nicely moved to your closed-source client code.

**Comment by Conturing** [comment]
That's good to hear. I will respond to all your annotations in the next few days.

As for the failed tests in `automata-serialization-dot`, I think these are related to the discussed changes in the DOTParser (added comperator).

File: api/src/main/java/net/automatalib/automata/visualization/MMCVisualizationHelper.java (hunk 2 above)
Comment (mtf90) [reply api/src/main/java/net/automatalib/automata/visualization/MMCVisualizationHelper.java:48]: Using `MMCAttrs.MEMBERSHIP` to associate an edge with a cluster is fine with me, but I think (DOT-specific) color markup should be client-specific (e.g. only *you* know that there are 14 clusters, etc.)

File: serialization/dot/src/test/java/net/automatalib/serialization/dot/DOTDeserializationTest.java (hunk 8 above)
Comment (mtf90) [reply serialization/dot/src/test/java/net/automatalib/serialization/dot/DOTDeserializationTest.java:65]: Ok, my bad then. I'll see if I can make any sense out of it.

File: api/src/main/java/net/automatalib/ts/modal/transitions/GroupMemberEdge.java (hunk 3 above)
Comment (mtf90) [reply api/src/main/java/net/automatalib/ts/modal/transitions/GroupMemberEdge.java:5]: I'd vote for removing this feature then for now

File: util/src/test/java/net/automatalib/util/ts/modal/regression/RegressionTests.java (hunk 11 above)
Comment (mtf90) [reply util/src/test/java/net/automatalib/util/ts/modal/regression/RegressionTests.java:175]: I'd argue that debugging is mostly done via an IDE that allows you to directly inspect in-memory objects rather than `System.out.println`-ing intermediate results. Would `LOGGER.debug()` be a viable output alternative? This would not require any file magic and you could still manually copy the terminal output to a file if needed.

**Comment by mtf90** [comment]
Ok, I'll look into it then with the serialization rework.

File: core/src/main/java/net/automatalib/ts/modal/MembershipMC.java [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+        boolean wasNotPresent = communicationAlphabet.add(symbol);
+
+        assert getInputAlphabet().contains(symbol) : "Communication alphabet needs to be a subset of alphabet";
+
+        return wasNotPresent;
+    }
+
+    public void removeCommunicationSymbol(I symbol) {
+        communicationAlphabet.remove(symbol);
+    }
```
Comment (mtf90) [inline core/src/main/java/net/automatalib/ts/modal/MembershipMC.java:70]: Is this functionality really needed? Currently, I don't see this function used. Normally, the (mutable) automaton types only support incremental operations and removing things (e.g. states) is realized via additional interfaces (e.g. `ShrinkableAutomaton`).

File: api/src/main/java/net/automatalib/ts/modal/ModalTransitionSystem.java (hunk 1 above)
Comment (mtf90) [reply api/src/main/java/net/automatalib/ts/modal/ModalTransitionSystem.java:33]: For most algorithms it makes sense to only consider finite MTSs. How about we add an `IndefiniteModalTransitionSystem` markup interface if at any point we (or others) want to model infinite systems. The `ModalTransitionSystem` interface could then be a finite refinement (similar to the `Graph` and `IndefiniteGraph` interfaces).

Comment (Conturing) [reply api/src/main/java/net/automatalib/ts/modal/ModalTransitionSystem.java:33]: I think that is a nice abstraction without any limitations on client side. But maybe we should enforce image-finiteness, because important theorems do not hold for unbounded transition sets (and these theorems may be implicitly assumed in client code).

File: api/src/main/java/net/automatalib/automata/visualization/MMCVisualizationHelper.java (hunk 2 above)
Comment (Conturing) [reply api/src/main/java/net/automatalib/automata/visualization/MMCVisualizationHelper.java:48]: Ok

File: api/src/main/java/net/automatalib/ts/modal/transitions/GroupMemberEdge.java (hunk 3 above)
Comment (Conturing) [reply api/src/main/java/net/automatalib/ts/modal/transitions/GroupMemberEdge.java:5]: Ok

File: util/src/test/java/net/automatalib/util/ts/modal/regression/RegressionTests.java (hunk 11 above)
Comment (Conturing) [reply util/src/test/java/net/automatalib/util/ts/modal/regression/RegressionTests.java:175]: I like specific tests with descriptive error information. Unfortunately, there is no support for such `assert`s on graphs. Therefore we either need to define our own asserts (and I have done that to some degree in the client code), or we need to compromise.

Dumping the graphs as dot files on the file system was never meant as a long term solution. Nevertheless, it proved really useful. Using a logger would be a cleaner solution, but it is also less practical.

File: core/src/main/java/net/automatalib/ts/modal/MembershipMC.java (hunk 13 above)
Comment (Conturing) [reply core/src/main/java/net/automatalib/ts/modal/MembershipMC.java:70]: Yes, it is. Inside a heavily-iterated loop we use a pruning method which removes unused labels. This is also relevant for following steps, so reordering would change the semantics. Copying the whole modal contract inside the loop is to expensive, as the graph uses to much memory. Therefore we need an implementation with dynamic mutators (adding/shrinking).

Extracting the functionality into an interface is a good idea.

File: util/src/test/java/net/automatalib/util/ts/modal/regression/RegressionTests.java (hunk 11 above)
Comment (Conturing) [reply util/src/test/java/net/automatalib/util/ts/modal/regression/RegressionTests.java:175]: I think the best course would be to implement custom `assert`s and then using only logging and no file-utils.

File: api/src/main/java/net/automatalib/ts/modal/ModalTransitionSystem.java (hunk 1 above)
Comment (mtf90) [reply api/src/main/java/net/automatalib/ts/modal/ModalTransitionSystem.java:33]: By making the default `ModalTransitionSystem` finite in states and input symbols and your code requiring this finite type in the parameters, everything should be fine with the properties of the theorems and most of your code can stay the way it currently is. My only issue was the lack of an indefinite version (if people want to model such an MTS).

File: util/src/test/java/net/automatalib/util/ts/modal/regression/RegressionTests.java (hunk 11 above)
Comment (mtf90) [reply util/src/test/java/net/automatalib/util/ts/modal/regression/RegressionTests.java:175]: I just realized this is test code and not library code :flushed:. I'm fine with having some intermediate checks/asserts as you suggested. Since most serialization code only requires some form of `OutputStream` or `Writer` we could use something like a `StringWriter` which stores the output in an in-memory `String` for future comparisons. This would be more convenient than a logger, but we wouldn't require filesystem access.

However, in this concrete case, you check the serialization of a modal conjunction, which mixes two aspects: The conjunction operation and the serialization of the result. From a unit-testing perspective, checking only the former would better fit the separation of concerns. You already do this with the assertion in the next line. Plus: I haven't seen code that actually reads the `phil3-actual` files anywhere, so you are currently not verifying the output anyway, right?

For now, I would still vote for removing this line, because it currently doesn't serve any purpose. For future (post-reintegrate) steps, we can add more asserts, possibly including in-memory serialization, ok?

File: core/src/main/java/net/automatalib/ts/modal/MembershipMC.java (hunk 13 above)
Comment (mtf90) [reply core/src/main/java/net/automatalib/ts/modal/MembershipMC.java:70]: Does this only concern alphabet symbols of the communication alphabet or the removal of states as well? (regarding a possible *shrinkable* interface)

Is it important to store this alphabet on the (M)MC instance or would it be possible/more universal to pass the communication-alphabet-to-consider as a method parameter (similar to how you can e.g. compute a shortest path using only the given set of input symbols)?

**Comment by mtf90** [comment]
One more general question (not directly related to this PR but the previous import): Is the `BidirectionalGraph` interface on the `AbstractCompactMTS` still required? Other than in the tests, I don't see it used.

And regarding the next steps: Once the last questions are resolved, I will probably merge this PR without any further changes, so that your fork can remain as-is and you have a version working with your current private project. I will then refactor things on the `mts` branch of AutomataLib so that you can gradually adjust/refactor your private code and see, if I broke anything essential. Once all lights are green, I'll merge the `mts` branch back into `develop`.

File: api/src/main/java/net/automatalib/ts/modal/ModalTransitionSystem.java (hunk 1 above)
Comment (Conturing) [reply api/src/main/java/net/automatalib/ts/modal/ModalTransitionSystem.java:33]: Yes, I got that. I just wanted to add that maybe an infinte-state ts without that restriction on the transitions may have no use case, as many formal proofs and concepts require it.

But since it does not effect our client code, it is completely up to you.

File: util/src/test/java/net/automatalib/util/ts/modal/regression/RegressionTests.java (hunk 11 above)
Comment (Conturing) [reply util/src/test/java/net/automatalib/util/ts/modal/regression/RegressionTests.java:175]: I like your proposition with the `StringWriter`. Could one automatically flush them in case of an failure?

As for the serialization: it is used to allow debugging. The assert fails with a message like "expected false to be true", which does not help finding the cause of an error.

File: core/src/main/java/net/automatalib/ts/modal/MembershipMC.java (hunk 13 above)
Comment (Conturing) [reply core/src/main/java/net/automatalib/ts/modal/MembershipMC.java:70]: Storing the alphabet separately is an option, but it would (probably) require changing many method signatures.

However, the "shrinked" alphabet is the communication alphabet. So when storing it separately it would always replace the intern communication alphabet, i.e. there is no semantic interpretation of the superset.

Other (shrinkable) methods may be used to further improve performance, but are currently not required.

**Comment by Conturing** [comment]
`BidirectionalGraph` is no longer needed :)

File: core/src/main/java/net/automatalib/ts/modal/MembershipMC.java (hunk 13 above)
Comment (mtf90) [reply core/src/main/java/net/automatalib/ts/modal/MembershipMC.java:70]: The main reason why I was worried is because with regular (input) alphabets, the symbol indexes are important for computing successors (e.g. each state has an array of length | \Sigma | which stores the transitions). So by removing input symbol this creates non-contiguous memory which we should prevent. However, so far it seems that this is only an additional property of an (M)MC and not related to any internal computations. For now, I'll leave it as-is, but we may discuss further possible cleanups in the future.

**Review by mtf90 — APPROVED** [review]
